### DISCUSSION TITLE
Single-letter prepositions in Czech

### DISCUSSION THREAD

**Discussion by sorashi** [typst/typst#591] [General]
In the Czech language, there is a typographical rule which states that single-letter prepositions shall be placed on the same line as the following word.
- [Source EN (automatic translation)][2].
- [Source CZ (original)][1].

This occurs very often, and for example Word takes care of this automatically when writing in Czech (by using a non-breaking space).

For LaTeX, the most popular solution is an external program from 1997 called [vlna][4] (wave), which traverses your source files and adds non-breakable spaces `~` wherever needed to conform to this rule.
This is an external binary and must therefore be run as a part of your build process.

This program can be imported as a [package][3], but only if you use XeTeX.
Otherwise, the package won't work.

I've always found this frustrating.
The job of typesetting systems should be to take care of the basic typographical rules in the author's stead.
Even when it comes to languages other than English that may use different typographical rules.

How should this be achieved in Typst?
- Is (or will be) the Typst scripting language powerful enough to write a module that transforms the entire document's content?
    - I read about the [query][5] function, but it seems unable to achieve this for now.
- If not, can this specific typographical localization rule be implemented in Typst itself?
- If not, then Typst is doomed to the same fate, where a program will be implemented externally, to add `~` wherever needed, producing source files riddled with `~`, and making it possible for authors to forget this build step before publishing their documents. Is that an alternative that is acceptable to Typst's philosophy and its (future) ecosystem?

For testing:

```
#set text(lang: "cs")

Ahoj. Tohle je test, který má za úkol vyzkoušet, zda-li Typst dokáže dodržet pravidlo, které je v jazykové příručce Ústavu pro jazyk český.

Ahoj. Tohle je test, který má za úkol vyzkoušet, zda-li Typst dokáže dodržet pravidlo, které je v~jazykové příručce Ústavu pro jazyk český.
```

![image](https://user-images.githubusercontent.com/6270283/230047141-3a85d6f2-fad5-4968-867b-fd55f67ded0b.png)

P.S. It may unintentionally sound passively aggressive, but I ask very humbly.
You are doing a great job and I really believe Typst can become a modern alternative to ancient typesetting systems.

[1]: https://prirucka.ujc.cas.cz/?id=880
[2]: https://prirucka-ujc-cas-cz.translate.goog/?id=880&_x_tr_sl=cs&_x_tr_tl=en&_x_tr_hl=cs&_x_tr_pto=wapp
[3]: https://ctan.org/pkg/xevlna?lang=en
[4]: https://petr.olsak.net/#:~:text=Program%20vlna
[5]: https://typst.app/docs/reference/meta/query/

**Comment by u-fischer** [comment]
> This program can be imported as a [package](https://ctan.org/pkg/xevlna?lang=en), but only if you use XeTeX.

For lua(la)tex there is the package luavlna https://ctan.org/pkg/luavlna

**Reply by Enivex** [reply]
Hey a LaTeX team spy 🙃

**Reply by sorashi** [reply]
The problem is, sometimes I cannot use XeTeX nor LuaTeX. I think it has something to do with PDF/A in combination with 8-bit font encoding.

```tex
\usepackage[a-2u]{pdfx}
\usepackage[czech]{babel}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
```

Errors (with XeTeX and LuaTeX, but it works with pdfLaTeX):

```
Latexmk: ====List of undefined refs and citations:
  Missing character: There is no Č (U+010C) in font ec-lmbx12!
  Missing character: There is no ž (U+017E) in font ec-lmr17!
  Missing character: There is no ř (U+0159) in font ec-lmbx12!
  Missing character: There is no ů (U+016F) in font ec-lmbx12!
  Missing character: There is no š (U+0161) in font ec-lmbx12!
  Missing character: There is no ů (U+016F) in font ec-lmbx12!
  Missing character: There is no ř (U+0159) in font ec-lmbx12!
 And 1090 more --- see log file.
```

Maybe I need to specify a different font? Either way, there is too much inconsistency and shenanigans here than really necessary. These problems could be solved preemptively in Typst.

**Reply by Enivex** [reply]
> The problem is, sometimes I cannot use XeTeX nor LuaTeX. I think it has something to do with PDF/A in combination with 8-bit font encoding.
> 
> ```
> \usepackage[a-2u]{pdfx}
> \usepackage[czech]{babel}
> \usepackage[T1]{fontenc}
> \usepackage[utf8]{inputenc}
> ```
> 
> Errors (with XeTeX and LuaTeX, but it works with pdfLaTeX):
> 
> ```
> Latexmk: ====List of undefined refs and citations:
>   Missing character: There is no Č (U+010C) in font ec-lmbx12!
>   Missing character: There is no ž (U+017E) in font ec-lmr17!
>   Missing character: There is no ř (U+0159) in font ec-lmbx12!
>   Missing character: There is no ů (U+016F) in font ec-lmbx12!
>   Missing character: There is no š (U+0161) in font ec-lmbx12!
>   Missing character: There is no ů (U+016F) in font ec-lmbx12!
>   Missing character: There is no ř (U+0159) in font ec-lmbx12!
>  And 1090 more --- see log file.
> ```
> 
> Maybe I need to specify a different font? Either way, there is too much inconsistency and shenanigans here than really necessary. These problems could be solved preemptively in Typst.

You shouldn't be using fontenc or inputenc with Luatex/Xetex, and even in pdftex  \usepackage[utf8]{inputenc} is redundant as of like 2 years ago.

**Reply by sorashi** [reply]
I see, thank you for the info, I'm going to look it up.

**Reply by u-fischer** [reply]
> The problem is, sometimes I cannot use XeTeX nor LuaTeX. I think it has something to do with PDF/A in combination with 8-bit font encoding. 

I don't think that this is the right place to correct a luaLaTeX document, ask better on tex.sx. But beside this: I doubt very much that typst has currently sufficient support for a standard like pdf/A-2u.

**Reply by sorashi** [reply]
> I don't think that this is the right place to correct a luaLaTeX document, ask better on tex.sx.

Of course, I didn't come here to discuss LaTeX :smile:, that's why I said I'm going to look it up.

> I doubt very much that typst has currently sufficient support for a standard like pdf/A-2u.

Typst definitely does not support everything I need at the moment, including but not limited to pdf/A-2u. I'm just giving ideas about what is required for correct localized typography in the future.

**Comment by laurmaedje** [comment]
This seems like something Typst should just implement natively.

**Reply by ararunaufc** [reply]
This should then be scoped to the line-breaking / hyphenation algorithms and interact with the (future) i18n infrastructure.

**Comment by sorashi** [comment]
I've played around with this a little bit and found this workaround. For anyone who would like the functionality at the moment.

```typst
#let reg = regex("\b([ksvzouaiKSVZOUAI])\b[ ]*(\S+)")
#show reg: element => (
  element.text.replace(reg, it => it.captures.join(sym.space.nobreak))
)
```

Of course, this uses two regexes, which might not be sustainable on a very large document.

**Reply by laurmaedje** [reply]
by the way, you can also store the full `regex("..")` into `reg` and do `show reg:`

**Reply by sorashi** [reply]
> by the way, you can also store the full `regex("..")` into `reg` and do `show reg:`

thanks, updated
